mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -8232,45 +9260,97 @@ struct ThreadwiseGemmRewritePattern
-    auto lbK = loopK.getBody();
-    b.setInsertionPointToStart(lbK);
-
-    auto loopM = b.create<AffineForOp>(loopK.getLoc(), 0, gemmAShape[2]);
-    auto lbM = loopM.getBody();
-    b.setInsertionPointToStart(lbM);
-
-    auto loopN = b.create<AffineForOp>(loc, 0, gemmBShape[2]);
-    auto lbN = loopN.getBody();
-    b.setInsertionPointToStart(lbN);
-
-    SmallVector<Value, 3> memIndicesKM;
-    extractForInductionVars({loopG, loopK, loopM}, &memIndicesKM);
-    auto gemmAKM = b.create<AffineLoadOp>(loc, gemmA, memIndicesKM);
-
-    SmallVector<Value, 3> memIndicesKN;
-    extractForInductionVars({loopG, loopK, loopN}, &memIndicesKN);
-    auto gemmBKN = b.create<AffineLoadOp>(loc, gemmB, memIndicesKN);
-
-    Value mul;
-    if (dataType.isa<IntegerType>())
-      mul = b.create<MulIOp>(loc, dataType, gemmAKM, gemmBKN);
-    else
-      mul = b.create<MulFOp>(loc, dataType, gemmAKM, gemmBKN);
-    SmallVector<Value, 3> memIndicesMN;
-    extractForInductionVars({loopG, loopM, loopN}, &memIndicesMN);
-    auto gemmCMN = b.create<AffineLoadOp>(loc, gemmC, memIndicesMN);
-
-    Value add;
-    if (dataType.isa<IntegerType>())
-      add = b.create<AddIOp>(loc, dataType, mul, gemmCMN);
-    else
-      add = b.create<AddFOp>(loc, dataType, mul, gemmCMN);
-    b.create<AffineStoreOp>(loc, add, gemmC, memIndicesMN);
+    assert(gemmAShape.size() == gemmBShape.size());
+    assert((gemmAShape.size() == 3) || (gemmAShape.size() == 4));
+    if (gemmAShape.size() == 3) {
+      // non-KPack path.
+      auto loopG = b.create<AffineForOp>(loc, 0, gemmAShape[0]);
+      auto lbG = loopG.getBody();
+      b.setInsertionPointToStart(lbG);
+
+      auto loopK = b.create<AffineForOp>(loc, 0, gemmAShape[1]);
+      auto lbK = loopK.getBody();
+      b.setInsertionPointToStart(lbK);
+
+      auto loopM = b.create<AffineForOp>(loopK.getLoc(), 0, gemmAShape[2]);
+      auto lbM = loopM.getBody();
+      b.setInsertionPointToStart(lbM);
+
+      auto loopN = b.create<AffineForOp>(loc, 0, gemmBShape[2]);
+      auto lbN = loopN.getBody();
+      b.setInsertionPointToStart(lbN);
+
+      SmallVector<Value, 3> memIndicesKM;
+      extractForInductionVars({loopG, loopK, loopM}, &memIndicesKM);
+      auto gemmAKM = b.create<AffineLoadOp>(loc, gemmA, memIndicesKM);
+
+      SmallVector<Value, 3> memIndicesKN;
+      extractForInductionVars({loopG, loopK, loopN}, &memIndicesKN);
+      auto gemmBKN = b.create<AffineLoadOp>(loc, gemmB, memIndicesKN);
+
+      Value mul;
+      if (dataType.isa<IntegerType>())
+        mul = b.create<MulIOp>(loc, dataType, gemmAKM, gemmBKN);
+      else
+        mul = b.create<MulFOp>(loc, dataType, gemmAKM, gemmBKN);
+      SmallVector<Value, 3> memIndicesMN;
+      extractForInductionVars({loopG, loopM, loopN}, &memIndicesMN);
+      auto gemmCMN = b.create<AffineLoadOp>(loc, gemmC, memIndicesMN);
+
+      Value add;
+      if (dataType.isa<IntegerType>())
+        add = b.create<AddIOp>(loc, dataType, mul, gemmCMN);
+      else
```

Comment:
Braces on all the if statements, please? It makes adding debugging info easier, and I'm tempted to ask we add a lint forbidding this sort of thing going forward.

Reply:
done